go/libraries/doltcore/sqle/binlogreplication/binlog_replica_applier.go

```diff
@@ -0,0 +1,997 @@
+			return nil
+		}
+		a.modifiedDatabases[tableMap.Database] = struct{}{}
+
+		rows, err := event.Rows(a.format, tableMap)
+		if err != nil {
+			return err
+		}
+
+		flags := rows.Flags
+		if flags&rowFlag_endOfStatement == rowFlag_endOfStatement {
+			// nothing to be done for end of statement; just clear the flag and move on
+			flags = flags &^ rowFlag_endOfStatement
+		}
+		if flags&rowFlag_noForeignKeyChecks == rowFlag_noForeignKeyChecks {
+			flags = flags &^ rowFlag_noForeignKeyChecks
+		}
+		if flags != 0 {
+			msg := fmt.Sprintf("unsupported binlog protocol message: UpdateRows event with unsupported flags '%x'", flags)
+			logger.Errorf(msg)
+			DoltBinlogReplicaController.setSqlError(mysql.ERUnknownError, msg)
+		}
+		schema, err := getTableSchema(ctx, engine, tableMap.Name, tableMap.Database)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf(" - Updated Rows (table: %s)", tableMap.Name)
+		for _, row := range rows.Rows {
+			identifyRow, err := parseRow(ctx, tableMap, schema, rows.IdentifyColumns, row.NullIdentifyColumns, row.Identify)
+			if err != nil {
+				return err
+			}
+			updatedRow, err := parseRow(ctx, tableMap, schema, rows.DataColumns, row.NullColumns, row.Data)
+			if err != nil {
+				return err
+			}
+			logger.Debugf("     - Identify: %v Data: %v ", sql.FormatRow(identifyRow), sql.FormatRow(updatedRow))
+
+			foreignKeyChecksDisabled := tableMap.Flags&rowFlag_noForeignKeyChecks > 0
+			writeSession, tableWriter, err := getTableWriter(ctx, engine, tableMap.Name, tableMap.Database, foreignKeyChecksDisabled)
+			if err != nil {
+				return err
+			}
+
+			err = tableWriter.Update(ctx, identifyRow, updatedRow)
+			if err != nil {
+				return err
+			}
+
+			err = closeWriteSession(ctx, engine, tableMap.Database, writeSession)
+			if err != nil {
+				return err
+			}
+		}
+
+	default:
+		// We can't access the bytes directly because these non-interface types in Vitess are not exposed.
+		// Having a Bytes() or Type() method on the Vitess interface would let us clean this up.
+		byteString := fmt.Sprintf("%v", event)
+		if strings.HasPrefix(byteString, "{[0 0 0 0 27 ") {
+			// Type 27 is a Heartbeat event. This event does not appear in the binary log. It's only sent over the
+			// network by a primary to a replica to let it know that the primary is still alive, and is only sent
+			// when the primary has no binlog events to send to replica servers.
+			// For more details, see: https://mariadb.com/kb/en/heartbeat_log_event/
+			logger.Debug("Received binlog event: Heartbeat")
+		} else {
+			return fmt.Errorf("received unknown event: %v", event)
+		}
+	}
+
+	if createCommit {
+		databasesToCommit := keys(a.modifiedDatabases)
+		if commitToAllDatabases {
+			databasesToCommit = getAllUserDatabaseNames(ctx, engine)
+		}
+		for _, database := range databasesToCommit {
+			executeQueryWithEngine(ctx, engine, "use `"+database+"`;")
+			executeQueryWithEngine(ctx, engine, "commit;")
```

Comment:
You're right that this could be simpler, although there are still cases with query events where we do need to commit explicitly. Thanks for pointing out that commit is a no-op if we've already flushed WriteSession changes to disk. I was able to simplify this so that row updates don't trigger a commit statement here, but we still do this for query events where we directly execute a statement.